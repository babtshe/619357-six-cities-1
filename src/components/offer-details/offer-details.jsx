import React, {PureComponent} from 'react';
import {propTypes} from './offer-details.props';
import {connect} from 'react-redux';
import {getOfferById, getReviews, getNearestOffers} from '../../redux/selectors';
import {fetchReviews} from '../../redux/operations';
import ReviewsList from '../reviews-list/reviews-list';
import OffersList from '../offers-list/offers-list';
import BookmarkButton from '../bookmark-button/bookmark-button';
import withActiveItem from '../../hocs/with-active-item/with-active-item';
import withToggleBookmark from '../../hocs/with-toggle-bookmark/with-toggle-bookmark';
import CityMap from '../city-map/city-map';

const ToggleBookmarkWrapped = withToggleBookmark(BookmarkButton);
const OffersListWrapped = withActiveItem(OffersList);

class OfferDetails extends PureComponent {
  constructor(props) {
    super(props);
    this.topElement = React.createRef();
    this.scrollToTop = this.scrollToTop.bind(this);
  }

  render() {
    if (!this.props.offer) {
      return null;
    }
    const {
      offer,
      reviews,
      nearestOffers,
    } = this.props;
    const {
      images,
      premium,
      name,
      rating,
      type,
      bedrooms,
      maxGuests,
      price,
      goods,
      host,
      description,
      bookmarked,
      id,
    } = offer;
    return (
      <main ref={this.topElement} className="page__main page__main--property">
        <section className="property">
          <div className="property__gallery-container container">
            <div className="property__gallery">
              {(images || []).map((image) => {
                return (
                  <div className="property__image-wrapper" key={image}>
                    <img className="property__image" src={image} alt="Photo studio"/>
                  </div>
                );
              })}
            </div>
          </div>
          <div className="property__container container">
            <div className="property__wrapper">
              {premium ?
                <div className="property__mark">
                  <span>Premium</span>
                </div> : ``
              }
              <div className="property__name-wrapper">
                <h1 className="property__name">
                  {name}
                </h1>
                <ToggleBookmarkWrapped
                  id={id}
                  bookmarked={bookmarked}
                />
              </div>
              <div className="property__rating rating">
                <div className="property__stars rating__stars">
                  <span style={{width: `${rating}%`}}></span>
                  <span className="visually-hidden">Rating</span>
                </div>
                <span className="property__rating-value rating__value">{rating / 20}</span>
              </div>
              <ul className="property__features">
                <li className="property__feature property__feature--entire">
                  {type === `house` ? `Entire place` : `${type.charAt(0).toUpperCase()}${type.slice(1)}`}

                </li>
                <li className="property__feature property__feature--bedrooms">
                  {bedrooms} Bedroom{(bedrooms === 1) ? `` : `s`}
                </li>
                <li className="property__feature property__feature--adults">
                  Max {maxGuests} adult{(maxGuests === 1) ? `` : `s`}
                </li>
              </ul>
              <div className="property__price">
                <b className="property__price-value">&euro;{price}</b>
                <span className="property__price-text">&nbsp;night</span>
              </div>
              <div className="property__inside">
                <h2 className="property__inside-title">What&apos;s inside</h2>
                <ul className="property__inside-list">
                  {(goods || []).map((item) => {
                    return (
                      <li className="property__inside-item" key={item}>
                        {item}
                      </li>
                    );
                  })}
                </ul>
              </div>
              <div className="property__host">
                <h2 className="property__host-title">Meet the host</h2>
                <div className="property__host-user user">
                  <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                    <img className="property__avatar user__avatar" src={host.avatar} width="74" height="74" alt="Host avatar"/>
                  </div>
                  <span className="property__user-name">
                    {host.name}
                  </span>
                  {host.pro ?
                    <span className="property__user-status">
                      Pro
                    </span> : ``
                  }
                </div>
                <div className="property__description">
                  <p className="property__text">
                    {description}
                  </p>
                </div>
              </div>
              <ReviewsList
                reviews={reviews}
                id={id}
              />
            </div>
          </div>
          <section className="property__map map">
            <CityMap
              locations={this.props.nearestOffers.reduce((acc, item) => {
                acc.push(item.location);
                return acc;
              }, [this.props.offer.location])}
              cityLocation={this.props.nearestOffers[0].city.location}
              zoom={nearestOffers[0].city.zoom}
              activeLocation = {this.props.offer.location}
            />
          </section>
        </section>
        <div className="container">
          <section className="near-places places">
            <h2 className="near-places__title">Other places in the neighbourhood</h2>
            <OffersListWrapped
              offers = {this.props.nearestOffers}
              classPrefix = {`near`}
            />
          </section>
        </div>
      </main>
    );
  }

  componentDidMount() {
    this.props.fetchReviews(this.props.offerId);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.offerId !== this.props.offerId) {
      this.props.fetchReviews(this.props.offerId);
      this.scrollToTop();
    }
  }

  scrollToTop() {
    this.topElement.current.scrollIntoView();
  }
}

OfferDetails.propTypes = propTypes;

const mapStateToProps = (state, ownProps) => ({
  offer: getOfferById(state, ownProps),
  reviews: getReviews(state),
  nearestOffers: getNearestOffers(state, ownProps),
});

const mapDispatchToProps = {
  fetchReviews: (id) => fetchReviews(id),
};

export {OfferDetails};

export default connect(mapStateToProps, mapDispatchToProps)(OfferDetails);
