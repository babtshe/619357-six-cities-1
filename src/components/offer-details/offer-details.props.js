import PropTypes from 'prop-types';
import {cardTypes} from '../offer-card/offer-card.props';

export const hostTypes = {
  name: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  pro: PropTypes.bool,
};

export const propTypes = {
  offer: PropTypes.shape(Object.assign({}, cardTypes, {
    images: PropTypes.arrayOf(PropTypes.string),
    bedrooms: PropTypes.number.isRequired,
    maxGuests: PropTypes.number.isRequired,
    goods: PropTypes.arrayOf(PropTypes.string),
    host: PropTypes.shape(hostTypes).isRequired,
  })).isRequired,
  onAddFavorites: PropTypes.func,
};
