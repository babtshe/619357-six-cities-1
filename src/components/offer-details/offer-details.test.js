import React from 'react';
import renderer from 'react-test-renderer';
import {OfferDetails} from './offer-details';

it(`OfferDetails should render correctly`, ()=>{
  const mockOffer = {
    name: `test1`,
    type: `apartment`,
    price: 1,
    rating: 1,
    image: `test2`,
    premium: true,
    id: 1,
    location: [52.3909553943508, 4.85309666406198],
    host: {
      name: `host`,
      id: 1,
      avatar: `test`,
    },
    goods: [`test`],
    images: [`test`],
    maxGuests: 1,
    bedrooms: 1,
  };
  const tree = renderer
  .create(
      <OfferDetails
        offer={mockOffer}
      />)
  .toJSON();

  expect(tree).toMatchSnapshot();
});
