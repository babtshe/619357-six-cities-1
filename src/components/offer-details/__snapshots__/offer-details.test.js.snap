// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferDetails should render correctly 1`] = `
<main
  className="page__main page__main--property"
>
  <section
    className="property"
  >
    <div
      className="property__gallery-container container"
    >
      <div
        className="property__gallery"
      >
        <div
          className="property__image-wrapper"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="test"
          />
        </div>
      </div>
    </div>
    <div
      className="property__container container"
    >
      <div
        className="property__wrapper"
      >
        <div
          className="property__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="property__name-wrapper"
        >
          <h1
            className="property__name"
          >
            test1
          </h1>
          <Memo(ConnectFunction)
            id={1}
          />
        </div>
        <div
          className="property__rating rating"
        >
          <div
            className="property__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "1%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
          <span
            className="property__rating-value rating__value"
          >
            0.05
          </span>
        </div>
        <ul
          className="property__features"
        >
          <li
            className="property__feature property__feature--entire"
          >
            Apartment
          </li>
          <li
            className="property__feature property__feature--bedrooms"
          >
            1
             Bedroom
            
          </li>
          <li
            className="property__feature property__feature--adults"
          >
            Max 
            1
             adult
            
          </li>
        </ul>
        <div
          className="property__price"
        >
          <b
            className="property__price-value"
          >
            €
            1
          </b>
          <span
            className="property__price-text"
          >
             night
          </span>
        </div>
        <div
          className="property__inside"
        >
          <h2
            className="property__inside-title"
          >
            What's inside
          </h2>
          <ul
            className="property__inside-list"
          >
            <li
              className="property__inside-item"
            >
              test
            </li>
          </ul>
        </div>
        <div
          className="property__host"
        >
          <h2
            className="property__host-title"
          >
            Meet the host
          </h2>
          <div
            className="property__host-user user"
          >
            <div
              className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
            >
              <img
                alt="Host avatar"
                className="property__avatar user__avatar"
                height="74"
                src="test"
                width="74"
              />
            </div>
            <span
              className="property__user-name"
            >
              host
            </span>
            
          </div>
          <div
            className="property__description"
          >
            <p
              className="property__text"
            />
          </div>
        </div>
        <Memo(ConnectFunction)
          id={1}
          reviews={Array []}
        />
      </div>
    </div>
    <section
      className="property__map map"
    >
      <CityMap
        activeLocation={
          Array [
            52.3909553943508,
            4.85309666406198,
          ]
        }
        cityLocation={
          Array [
            52.3909553943508,
            4.85309666406198,
          ]
        }
        locations={
          Array [
            Array [
              52.3909553943508,
              4.85309666406198,
            ],
            Array [
              52.3909553943508,
              4.85309666406198,
            ],
          ]
        }
      />
    </section>
  </section>
  <div
    className="container"
  >
    <section
      className="near-places places"
    >
      <h2
        className="near-places__title"
      >
        Other places in the neighbourhood
      </h2>
      <WithActiveItem
        classPrefix="near"
        offers={
          Array [
            Object {
              "bedrooms": 1,
              "city": Object {
                "location": Array [
                  52.3909553943508,
                  4.85309666406198,
                ],
              },
              "goods": Array [
                "test",
              ],
              "host": Object {
                "avatar": "test",
                "id": 1,
                "name": "host",
              },
              "id": 1,
              "image": "test2",
              "images": Array [
                "test",
              ],
              "location": Array [
                52.3909553943508,
                4.85309666406198,
              ],
              "maxGuests": 1,
              "name": "test1",
              "premium": true,
              "price": 1,
              "rating": 1,
              "type": "apartment",
            },
          ]
        }
      />
    </section>
  </div>
</main>
`;
